src/briefcase/platforms/android/apk.py

```diff
@@ -1 +1,268 @@
+
+from briefcase.commands import (
+    BuildCommand,
+    CreateCommand,
+    PackageCommand,
+    PublishCommand,
+    RunCommand,
+    UpdateCommand,
+)
+from briefcase.config import BaseConfig
+from briefcase.exceptions import BriefcaseCommandError, NetworkFailure
+from briefcase.integrations.adb import (
+    no_or_wrong_device_message,
+    force_stop_app,
+    install_apk,
+    start_app,
+)
+
+
+class ApkMixin:
+    output_format = "apk"
+    platform = "android"
+    dot_briefcase = Path.home() / ".briefcase"
+
+    def binary_path(self, app):
+        return (
+            self.platform_path
+            / app.formal_name
+            / "app"
+            / "build"
+            / "outputs"
+            / "apk"
+            / "debug"
+            / "app-debug.apk"
+        )
+
+    def distribution_path(self, app):
+        return self.binary_path(app)
+
+    @property
+    def sdk_path(self):
+        return self.dot_briefcase / "tools" / "android_sdk"
+
+    @property
+    def sdk_url(self):
+        """The Android SDK URL appropriate to the current operating system."""
+        # The URLs described by the pattern below have existed since
+        # approximately 2017, and the code they download has a built-in
+        # updater. I hope they will work for many years.
+        return "https://dl.google.com/android/repository/" + (
+            "sdk-tools-{os}-4333796.zip".format(os=self.host_os.lower())
+        )
+
+    def verify_python_version(self):
+        if self.python_version_tag != "3.7":
+            raise BriefcaseCommandError(
+                """\
+Found Python version {self.python_version_tag}. Android packaging currently
+requires Python 3.7.""".format(
+                    self=self
+                )
+            )
+
+    def verify_sdk(self):
+        # If the SDK tools don't exist or aren't executable, install the SDK.
+        # TODO: Reconsider this for Windows.
+        tools_path = self.sdk_path / "tools" / "bin"
+        tools_ok = (
+            tools_path.exists()
+            and all(
+                [
+                    S_IMODE(tool.stat().st_mode) == 0o755
+                    for tool in tools_path.glob("*")
+                ]
+            )
+            and (tools_path / "sdkmanager").exists()
+        )
+        if tools_ok:
+            return
+
+        print("Setting up Android SDK...")
+        try:
+            sdk_zip_path = self.download_url(
+                url=self.sdk_url,
+                download_path=self.dot_briefcase / "tools",
+            )
+        except requests_exceptions.ConnectionError:
+            raise NetworkFailure("download Android SDK")
+        try:
+            with ZipFile(str(sdk_zip_path)) as sdk_zip:
+                sdk_zip.extractall(path=str(self.sdk_path))
+        except BadZipFile:
+            raise BriefcaseCommandError(
+                """\
+Invalid ZIP file found at {sdk_zip_path}
+
+Partial download? Remove it, then try again.""".format(
```

Comment:
This error message needs to be reworded a little. It's technically correct, but not especially clear for non-expert readers.

Reply:
From gitter conversation, the new plan is:

```
Unable to unpack Android SDK ZIP file. The download may have been interrupted or corrupted.

Delete {sdk_zip_path} and run briefcase again.
```